src/cpp/modules/finite/finite.h

```diff
@@ -63,51 +93,41 @@ class FiniteDifferencesEngine
     /// <param name="input_size">Input data size (1 for scalar)</param>
-    /// <param name="output">Pointer to where 'func' should output data (scalar or vector)</param>
     /// <param name="output_size">Size of 'func' output data</param>
     /// <param name="result">Pointer to where resultant Jacobian should go.
     ///		Will be stored as a vector(input_size * output_size).
     ///		Will store in format foreach (input) { foreach (output) {} }</param>
-    /// <param name="delta">The difference to use in finite differentiation</param>
     void finite_differences(std::function<void(T*, T*)> func,
-        T* input, int input_size, T* output, int output_size,
-        T* result, T delta = FINITE_DIFFERENCES_DEFAULT_DELTA)
-    {
-        func(input, output);
-        finite_differences_continue(func, input, input_size, output, output_size, result, delta);
-    }
-
-    /// <summary>Approximately differentiate a function using finite differences.
-    ///     This variation expects 'func(input)' to be pre-computed.
-    ///     It can be used to continuously, part-by-part compute the gradient of one function.</summary>
-    /// <param name="func">Function to be differentiated.
-    ///		Should accept 2 pointers as arguments (one input, one output).
-    ///		Each output will be differentiated with respect to each input.</param>
-    /// <param name="input">Pointer to input data (scalar or vector)</param>
-    /// <param name="input_size">Input data size (1 for scalar)</param>
-    /// <param name="output">Pointer to pre-computed 'func(input)'s output data (scalar or vector)</param>
-    /// <param name="output_size">Size of 'func' output data</param>
-    /// <param name="result">Pointer to where resultant Jacobian should go.
-    ///		Will be stored as a vector(input_size * output_size).
-    ///		Will store in format foreach (input) { foreach (output) {} }</param>
-    /// <param name="delta">The difference to use in finite differentiation</param>
-    void finite_differences_continue(std::function<void(T*, T*)> func,
-        T* input, int input_size, const T* output, int output_size,
-        T* result, T delta = FINITE_DIFFERENCES_DEFAULT_DELTA)
+        T* input, int input_size, int output_size,
+        T* result)
     {
+        volatile T tmp_f, tmp_b;
         for (int i = 0; i < input_size; i++)
         {
             T originalInput = input[i];
-            input[i] += delta;
-            func(input, tmp_output.data());
-            div_vec(sub_vec(tmp_output.data(), output, output_size), output_size, delta);
-            vec_ins(&result[output_size * i], tmp_output.data(), output_size);
+            T delta = step(originalInput);
+            tmp_b = originalInput - delta;
+            T dx = delta * 2;
+            tmp_f = tmp_b + dx;
+            // adjusting dx so that (tmp_b + dx) - tmp_b == delta
+            dx = tmp_f - tmp_b;
+            if (dx < delta * 0.02)
```

Comment:
Did you consider 0.5?  Does it generate more errors?
I'm kind of keen on powers of two for this constant, even though I realize it's somewhat arbitrary.

Reply:
No, not initially.

No it doesn't - neither of the two generated any errors so far.

To think of it, for `dx` to be less that `0.5 * delta`, while without round-off errors it would be `2 * delta`, it, probably, must be zero, because that means that the round-off error of the last addition was greater than `0.5 * dx`, so not a single bit of `dx`'s significand was added correctly.

So, yeah, let's make it 0.5.